File: js/src/lru-cache.test.ts
import { describe, expect, it } from "vitest";
import { LRUCache } from "./lru-cache";

describe("LRUCache", () => {
  it("should store and retrieve values", () => {
    const cache = new LRUCache<string, number>();
    cache.set("a", 1);
    expect(cache.get("a")).toBe(1);
  });

  it("should return undefined for missing keys", () => {
    const cache = new LRUCache<string, number>();
    expect(cache.get("missing")).toBeUndefined();
  });

  it("should respect max size when specified", () => {
    const cache = new LRUCache<string, number>({ max: 2 });
    cache.set("a", 1);
    cache.set("b", 2);
    cache.set("c", 3);
    expect(cache.get("a")).toBeUndefined();
    expect(cache.get("b")).toBe(2);
    expect(cache.get("c")).toBe(3);
  });

  it("should grow unbounded when no max size specified", () => {
    const cache = new LRUCache<number, number>();
    // Add many items.
    for (let i = 0; i < 1000; i++) {
      cache.set(i, i);
    }
    // Should keep all items.
    for (let i = 0; i < 1000; i++) {
      expect(cache.get(i)).toBe(i);
    }
  });

  it("should refresh items on get", () => {
    const cache = new LRUCache<string, number>({ max: 2 });
    cache.set("a", 1);
    cache.set("b", 2);
    cache.get("a"); // refresh "a"
    cache.set("c", 3);
    expect(cache.get("a")).toBe(1);
    expect(cache.get("b")).toBeUndefined();
    expect(cache.get("c")).toBe(3);
  });

  it("should update existing keys", () => {
    const cache = new LRUCache<string, number>();
    cache.set("a", 1);
    cache.set("a", 2);
    expect(cache.get("a")).toBe(2);
  });

  it("should clear all items", () => {
    const cache = new LRUCache<string, number>();
    cache.set("a", 1);
    cache.set("b", 2);
    cache.clear();
    expect(cache.get("a")).toBeUndefined();
    expect(cache.get("b")).toBeUndefined();
  });
});

File: py/src/braintrust/lru_cache_test.py
import unittest

from .lru_cache import LRUCache


class TestLRUCache(unittest.TestCase):
    def test_store_and_retrieve_values(self):
        """Test storing and retrieving values."""
        cache = LRUCache[str, int]()
        cache.set("a", 1)
        self.assertEqual(cache.get("a"), 1)

    def test_missing_keys(self):
        """Test raising KeyError for missing keys."""
        cache = LRUCache[str, int]()
        with self.assertRaises(KeyError):
            cache.get("missing")

    def test_respect_max_size(self):
        """Test respecting max size when specified."""
        cache = LRUCache[str, int](max_size=2)
        cache.set("a", 1)
        cache.set("b", 2)
        cache.set("c", 3)
        with self.assertRaises(KeyError):
            cache.get("a")
        self.assertEqual(cache.get("b"), 2)
        self.assertEqual(cache.get("c"), 3)

    def test_grow_unbounded(self):
        """Test growing unbounded when no max size specified."""
        cache = LRUCache[int, int]()
        # Add many items.
        for i in range(1000):
            cache.set(i, i)
        # Should keep all items.
        for i in range(1000):
            self.assertEqual(cache.get(i), i)

    def test_refresh_items_on_get(self):
        """Test refreshing items on get."""
        cache = LRUCache[str, int](max_size=2)
        cache.set("a", 1)
        cache.set("b", 2)
        cache.get("a")  # refresh "a"
        cache.set("c", 3)
        self.assertEqual(cache.get("a"), 1)
        with self.assertRaises(KeyError):
            cache.get("b")
        self.assertEqual(cache.get("c"), 3)

    def test_update_existing_keys(self):
        """Test updating existing keys."""
        cache = LRUCache[str, int]()
        cache.set("a", 1)
        cache.set("a", 2)
        self.assertEqual(cache.get("a"), 2)

    def test_clear_all_items(self):
        """Test clearing all items."""
        cache = LRUCache[str, int]()
        cache.set("a", 1)
        cache.set("b", 2)
        cache.clear()
        with self.assertRaises(KeyError):
            cache.get("a")
        with self.assertRaises(KeyError):
            cache.get("b")


if __name__ == "__main__":
    unittest.main()
